### PR TITLE
Nested username attribute in DefaultOAuth2User

### PR DESCRIPTION
**Pull request by ahmd-nabil**
ISSUE gh-14186

### CODE REVIEW

File: oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/userinfo/DefaultOAuth2UserService.java
Comment (jzheaux) [inline oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/userinfo/DefaultOAuth2UserService.java:182]: Please move this to be a member variable as it doesn't need to be re-constructed each time.

File: oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/userinfo/DefaultOAuth2UserService.java
Comment (jzheaux) [inline oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/userinfo/DefaultOAuth2UserService.java:183]: I think it's important to rely on the developer to specify the correct expression. Are you able to simply pass in `userNameAttributeName`?

File: oauth2/oauth2-client/src/test/java/org/springframework/security/oauth2/client/userinfo/DefaultOAuth2UserServiceTests.java
Comment (jzheaux) [inline oauth2/oauth2-client/src/test/java/org/springframework/security/oauth2/client/userinfo/DefaultOAuth2UserServiceTests.java:165]: I believe the reason you are having trouble with the simple dot syntax is since `user-name` is not a valid SpEL accessor. 

Either change the payload to use `username` for the property or use `user['user-name']` for the expression.

File: oauth2/oauth2-core/src/main/java/org/springframework/security/oauth2/core/user/DefaultOAuth2User.java
Comment (jzheaux) [inline oauth2/oauth2-core/src/main/java/org/springframework/security/oauth2/core/user/DefaultOAuth2User.java:68]: Please add a description here as to what is the preferred alternative

File: oauth2/oauth2-core/src/main/java/org/springframework/security/oauth2/core/user/DefaultOAuth2User.java
Comment (jzheaux) [inline oauth2/oauth2-core/src/main/java/org/springframework/security/oauth2/core/user/DefaultOAuth2User.java:70]: Can this constructor be changed to call the new constructor? This may simplify future maintenance.

File: oauth2/oauth2-client/src/test/java/org/springframework/security/oauth2/client/jackson2/OAuth2AuthenticationTokenMixinTests.java
Comment (jzheaux) [inline oauth2/oauth2-client/src/test/java/org/springframework/security/oauth2/client/jackson2/OAuth2AuthenticationTokenMixinTests.java:197]: I don't think this is representative of how the JSON will look (no more `nameAttributeKey` property). 

The tests as a whole should confirm two things:

1. Values serialized with no `name` property (given that it is new) should still deserialize correctly
2. Values with no `nameAttributeKey` property (given that it is being removed) should still deserialize correctly

I believe this means that there will need to be a change to `DefaultOAuth2UserMixin`; it may need to use a deserializer, like `ClientRegistrationMixin` does. If it contains `nameAttributeKey`, use the deprecated constructor; otherwise, use the new constructor.

File: oauth2/oauth2-core/src/main/java/org/springframework/security/oauth2/core/user/DefaultOAuth2User.java
Comment (jzheaux) [inline oauth2/oauth2-core/src/main/java/org/springframework/security/oauth2/core/user/DefaultOAuth2User.java:58]: This should no longer be necessary

File: oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/userinfo/DefaultOAuth2UserService.java
Comment (jzheaux) [inline oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/userinfo/DefaultOAuth2UserService.java:144]: I appreciate this cleanup. So that the feature commit has only the changes needed for the feature, will you please place these other refactors into a separate, earlier commit in the PR?

**Review by jzheaux — CHANGES_REQUESTED** [review]
Thanks, @ahmd-nabil! I've left some feedback inline.

File: oauth2/oauth2-core/src/main/java/org/springframework/security/oauth2/core/user/DefaultOAuth2User.java
Comment (ahmd-nabil) [reply oauth2/oauth2-core/src/main/java/org/springframework/security/oauth2/core/user/DefaultOAuth2User.java:70]: @jzheaux, doing so caused `NullPointerException`s, because I am getting the `name` out of `attributes` map before checking for nulls in the new constructor.

File: oauth2/oauth2-client/src/test/java/org/springframework/security/oauth2/client/jackson2/OAuth2AuthenticationTokenMixinTests.java
Comment (ahmd-nabil) [reply oauth2/oauth2-client/src/test/java/org/springframework/security/oauth2/client/jackson2/OAuth2AuthenticationTokenMixinTests.java:197]: @jzheaux, here after implementing deserializer for `DefaultOAuth2UserMixin`, I had to implement another deserializer `DefaultOidcUserMixin`. seems reasonable ?? just checking if I am in the right direction

Comment (ahmd-nabil) [reply oauth2/oauth2-client/src/test/java/org/springframework/security/oauth2/client/jackson2/OAuth2AuthenticationTokenMixinTests.java:197]: @jzheaux, just to make sure I am going the right direction, working on this I had to create 2 deserializers one for `DefaultOAuth2UserMixin` and another for `DefaultOidcUserMixin` seems reasonable ?

File: oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/userinfo/DefaultOAuth2UserService.java
Comment (ahmd-nabil) [reply oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/userinfo/DefaultOAuth2UserService.java:145]: I will do so once everything else is approved

File: oauth2/oauth2-core/src/main/java/org/springframework/security/oauth2/core/user/DefaultOAuth2User.java
Comment (ahmd-nabil) [reply oauth2/oauth2-core/src/main/java/org/springframework/security/oauth2/core/user/DefaultOAuth2User.java:70]: another solution would be by introducing an `init` method to encapsulate common logic, but will require all `authorities`, `attributes`, and `name` to be **non final**.

init method would look something like
```
private void init(Map<String, Object> attributes, Collection<? extends GrantedAuthority> authorities,
					  String name) {
        this.authorities = (authorities != null)
		        ? Collections.unmodifiableSet(new LinkedHashSet<>(this.sortAuthorities(authorities)))
		        : Collections.unmodifiableSet(new LinkedHashSet<>(AuthorityUtils.NO_AUTHORITIES));
        this.attributes = Collections.unmodifiableMap(new LinkedHashMap<>(attributes));
        this.name = name;
}
```
old constructor
```
public DefaultOAuth2User(Collection<? extends GrantedAuthority> authorities, Map<String, Object> attributes,
			String nameAttributeKey) {
        Assert.notEmpty(attributes, "attributes cannot be empty");
        Assert.hasText(nameAttributeKey, "nameAttributeKey cannot be empty");
        if (!attributes.containsKey(nameAttributeKey)) {
	        throw new IllegalArgumentException("Missing attribute '" + nameAttributeKey + "' in attributes");
        }
        init(attributes, authorities,(String) attributes.get(nameAttributeKey));
}
```
and new constructor
```
public DefaultOAuth2User(Map<String, Object> attributes, Collection<? extends GrantedAuthority> authorities,
			String name) {
        Assert.notEmpty(attributes, "attributes cannot be empty");
        Assert.hasText(name, "name cannot be empty");
        init(attributes, authorities, name);
}
```

**Comment by ahmd-nabil** [comment]
Hi @jzheaux. sorry for rushing you. But, I 've been waiting for a review for a while, so I am just making sure if PR is still needed. whenever you have the time I would appreciate your review :)

**Comment by jzheaux** [comment]
Thanks for the updates, @ahmd-nabil. Some of your recent changes caused me to see that there might not be a way to remain backward compatible. For example, an application using `user-name-attribute: user-name` will break when upgrading.

Further, the JWT spec does not restrict characters for custom claim names, meaning that it is technically possible for `user.username` to be a valid claim name by itself. This makes it rather tricky to know the user's intent by reading the property value.

Given that, I think we might need to take a different route, and hopefully one that is less invasive. To that end, I've pushed a change to [my fork](https://github.com/spring-projects/spring-security/compare/main...jzheaux:spring-security:gh-14186) that takes a different approach. If you like it, you are welcome to merge it to your branch.

**Comment by ahmd-nabil** [comment]
thanks @jzheaux merged. and as always thanks for your guidance :)

**Comment by jzheaux** [comment]
Great, thanks @ahmd-nabil. Are you able to also do the same for `DefaultReactiveOAuth2UserService` and the associated tests? 

If not, I can add that as a polish. Either way, I think we're ready to separate the existing code cleanup into a separate commit and squash the other changes into a single commit.

**Comment by ahmd-nabil** [comment]
TY @jzheaux, I've separated the commit into 2 of them (I had to copy paste a lot to do that, I wonder if there is a better way).
I am not that good with project reactor but I will give it a try.

**Comment by jzheaux** [comment]
Thanks, @ahmd-nabil! This is now merged into `main`.

**Pull request merged**
